engine/apps/alerts/tasks/check_escalation_finished.py

```diff
@@ -1,48 +1,130 @@
+    """This exception is raised when an alert group's escalation policy did not execute execute properly for some reason"""
+
+
+def send_alert_group_escalation_auditor_task_heartbeat() -> None:
+    heartbeat_url = settings.ALERT_GROUP_ESCALATION_AUDITOR_CELERY_TASK_HEARTBEAT_URL
+    if heartbeat_url:
+        task_logger.info(f"Sending heartbeat to configured URL: {heartbeat_url}")
+        requests.get(heartbeat_url).raise_for_status()
+        task_logger.info(f"Heartbeat successfully sent to {heartbeat_url}")
+    else:
+        task_logger.info(f"Skipping sending heartbeat as no heartbeat URL is configured")
+
+
+def audit_alert_group_escalation(alert_group: "AlertGroup") -> None:
+    escalation_snapshot = alert_group.escalation_snapshot
+    alert_group_id = alert_group.id
+    base_msg = f"Alert group {alert_group_id}"
+
+    if not escalation_snapshot:
+        raise AlertGroupEscalationPolicyExecutionAuditException(
+            f"{base_msg} does not have an escalation snapshot associated with it, this should never occur"
+        )
+    task_logger.info(f"{base_msg} has an escalation snapshot associated with it, auditing if it executed properly")
+
+    escalation_policies_snapshots = escalation_snapshot.escalation_policies_snapshots
+
+    if not escalation_policies_snapshots:
+        # TODO: in what case is this valid?
+        task_logger.info(
+            f"{base_msg}'s escalation snapshot has an empty escalation_policies_snapshots, skipping further validation"
+        )
+        return
+    task_logger.info(
+        f"{base_msg}'s escalation snapshot has a populated escalation_policies_snapshots, continuing validation"
+    )
+
+    if escalation_snapshot.next_step_eta_is_valid() is False:
+        # TODO: is a next_step_eta value of None actually valid here, or should we raise an error?
+        raise AlertGroupEscalationPolicyExecutionAuditException(
+            f"{base_msg}'s escalation snapshot does not have a valid next_step_eta: {escalation_snapshot.next_step_eta}"
+        )
+    task_logger.info(f"{base_msg}'s escalation snapshot has a valid next_step_eta: {escalation_snapshot.next_step_eta}")
+
+    executed_escalation_policy_snapshots = escalation_snapshot.executed_escalation_policy_snapshots
+    num_of_executed_escalation_policy_snapshots = len(executed_escalation_policy_snapshots)
+
+    if num_of_executed_escalation_policy_snapshots == 0:
+        task_logger.info(
+            f"{base_msg}'s escalation snapshot does not have any executed escalation policies, skipping further validation"
+        )
+        return
+    task_logger.info(
+        f"{base_msg}'s escalation snapshot has {num_of_executed_escalation_policy_snapshots} executed escalation policies"
+    )
+
+    for executed_escalation_policy_snapshot in executed_escalation_policy_snapshots:
+        escalation_policy_id = executed_escalation_policy_snapshot.id
+
+        # TODO: is it valid to only check for the finished log record type here?
```

Comment:
question here

Reply:
We create this log if all escalation steps were executed or escalation chain didn't have any steps. As far as I know we don't create it only on step `Resolve incident automatically`, but in this case alert group would have status `resolved`, so it doesn't count.

Reply:
got it 👍

so it should be okay to just check for this log record type? I wasn't sure if I might also want to look for the "pending" type

Reply:
What do you mean by "pending type"?

I think for all successfully finished escalations this log should exist. 
But there can be a case, when escalation was finished, this log was created, and then user started escalation again (for example, resolve/unresolve). Seems like in this case it is possible to have the log record `TYPE_ESCALATION_FINISHED` and unfinished escalation at the same time 🤔

Reply:
> What do you mean by "pending type"?

ah sorry, ignore this. I thought there was a `AlertGroupLogRecord.TYPE_ESCALATION_PENDING` log record type, doesn't look like there is.

> But there can be a case, when escalation was finished, this log was created, and then user started escalation again (for example, resolve/unresolve). Seems like in this case it is possible to have the log record TYPE_ESCALATION_FINISHED and unfinished escalation at the same time

do you have any suggestions for the query in `has_finished_log_record` [method](https://github.com/grafana/oncall/pull/1266/files#diff-167ec80d5717a06a85c122857926e50b393abfdcae1fe511ca83ac00f256e23fR108-R113) on how to handle this edge case?

Reply:
I don't have any suggestions how to do it in one query 😞  The main thing is that we should check if escalation wasn't triggered again after this log, like there are no `TYPE_ESCALATION_TRIGGERED` or `TYPE_ESCALATION_FAILED` logs after `TYPE_ESCALATION_FINISHED` log